File: packages/react/src/trace-context.ts
export const TRACEPARENT_META_NAME = "traceparent";
export const TRACESTATE_META_NAME = "tracestate";

const TRACEPARENT_PATTERN =
  /^([0-9a-f]{2})-([0-9a-f]{32})-([0-9a-f]{16})-([0-9a-f]{2})$/;
const ALL_ZERO_TRACE_ID = "0".repeat(32);
const ALL_ZERO_SPAN_ID = "0".repeat(16);

export type DocumentTraceContext = {
  traceparent?: string;
  tracestate?: string;
};

function normalizeTraceParent(value: string | null | undefined): string {
  const normalized = value?.trim().toLowerCase() ?? "";
  const match = TRACEPARENT_PATTERN.exec(normalized);
  if (
    !match ||
    match[1] === "ff" ||
    match[2] === ALL_ZERO_TRACE_ID ||
    match[3] === ALL_ZERO_SPAN_ID
  ) {
    return "";
  }
  return normalized;
}

function normalizeTraceState(value: string | null | undefined): string {
  const normalized = value?.trim() ?? "";
  if (!normalized || normalized.length > 512) return "";
  for (let i = 0; i < normalized.length; i += 1) {
    const code = normalized.charCodeAt(i);
    if (code < 0x20 || code > 0x7e || code === 0x22 || code === 0x3c || code === 0x3e) {
      return "";
    }
  }
  return normalized;
}

function metaContent(target: Document, name: string): string | null {
  return target.querySelector(`meta[name="${name}"]`)?.getAttribute("content") ?? null;
}

/** Read the HTML document's W3C trace context so fetches can nest under the page request. */
export function getDocumentTraceContext(
  doc?: Document,
): DocumentTraceContext {
  const target = doc ?? (typeof document === "undefined" ? undefined : document);
  if (!target) return {};
  const traceparent = normalizeTraceParent(metaContent(target, TRACEPARENT_META_NAME));
  if (!traceparent) return {};
  const tracestate = normalizeTraceState(metaContent(target, TRACESTATE_META_NAME));
  return tracestate ? { traceparent, tracestate } : { traceparent };
}

/** Headers that continue the document request's W3C parent on a same-origin fetch. */
export function documentTraceHeaders(doc?: Document): Record<string, string> {
  const context = getDocumentTraceContext(doc);
  const headers: Record<string, string> = {};
  if (context.traceparent) headers.traceparent = context.traceparent;
  if (context.tracestate) headers.tracestate = context.tracestate;
  return headers;
}

File: document/tracecontext.go
package document

import (
	"regexp"
	"strings"
	"unicode"
)

// W3C Trace Context header names. Hosts such as gbhost inject these on the
// document request; the HTML document then exposes them to JavaScript so
// same-origin fetches can continue the page request instead of starting a
// new root.
const (
	TraceParentHeader = "traceparent"
	TraceStateHeader  = "tracestate"
	TraceParentMeta   = "traceparent"
	TraceStateMeta    = "tracestate"
)

var (
	traceParentPattern = regexp.MustCompile(`^([0-9a-f]{2})-([0-9a-f]{32})-([0-9a-f]{16})-([0-9a-f]{2})$`)
	allZeroTraceID     = strings.Repeat("0", 32)
	allZeroSpanID      = strings.Repeat("0", 16)
)

// NormalizeTraceParent returns a lowercase W3C traceparent, or empty when the
// value is missing or invalid. Invalid values are dropped so they cannot be
// reflected into HTML or reused as a parent.
func NormalizeTraceParent(value string) string {
	normalized := strings.ToLower(strings.TrimSpace(value))
	match := traceParentPattern.FindStringSubmatch(normalized)
	if match == nil || match[1] == "ff" || match[2] == allZeroTraceID || match[3] == allZeroSpanID {
		return ""
	}
	return normalized
}

// NormalizeTraceState returns a W3C tracestate value, or empty when it is
// missing or not safe to copy into HTML.
func NormalizeTraceState(value string) string {
	normalized := strings.TrimSpace(value)
	if normalized == "" || len(normalized) > 512 {
		return ""
	}
	for _, r := range normalized {
		if r < 0x20 || r > 0x7e || r == '"' || r == '<' || r == '>' {
			return ""
		}
		if unicode.IsControl(r) {
			return ""
		}
	}
	return normalized
}
